File: gatling-test-code/src/gatling/kotlin/requestClass/RequestAuth.kt
package requestClass

import java.security.MessageDigest
import java.util.*
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import java.time.LocalDateTime


class RequestAuth(
  private val method: String,
  private val path: String,
  queryParam: Map<String, String>? = null,
  requestBody: String? = null
) {
  private val date: String = LocalDateTime.now().toString()
  private var md5: String = calcMd5(requestBody)
  private val nonce: String = createUuid()
  private var url: String = urlToSign(queryParam)
  private var signature: String = getSign()

  companion object {
    const val url: String = "#{url}"

    //the header required by every api
    val sentHeader: Map<String, String> = mapOf(
      "Content-MD5" to "#{md5}",
      "X-Ca-Nonce" to "#{nonce}",
      "X-Ca-Signature" to "#{signature}",
      "Content-Type" to "application/json",
      "Date" to "#{Date}",
      "X-Ca-Signature-Headers" to "x-ca-key,x-ca-nonce,x-ca-signaturemethod",
      "X-Ca-Key" to System.getenv("APP_KEY"),
      "X-Ca-SignatureMethod" to "HmacSHA256"
    )

    //
    fun requestFeeder(method: String, path: String): Iterator<Map<String, String>> {
      return generateSequence {
        val requestAuth = RequestAuth(method, path)
        requestAuth.buildHeaderMap()
      }.iterator()
    }

    //request feeder for api with request body
    fun requestFeeder(
      method: String, path: String, requestBody: String
    ): Iterator<Map<String, String>> {
      return generateSequence {
        val requestAuth = RequestAuth(method, path, requestBody = requestBody)
        requestAuth.buildHeaderMap()
      }.iterator()
    }

    //request feeder for api with query parameter
    fun requestFeeder(
      method: String, path: String, queryParam: Map<String, String>
    ): Iterator<Map<String, String>> {
      return generateSequence {
        val requestAuth = RequestAuth(method, path, queryParam = queryParam)
        requestAuth.buildHeaderMap()
      }.iterator()
    }
  }

  fun buildHeaderMap(): Map<String, String> {
    return mapOf(
      "url" to url,
      "md5" to md5,
      "nonce" to nonce,
      "signature" to signature,
      "Date" to date,
    )
  }

  private fun createUuid(): String {
    return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(Regex("[xy]")) { c ->
      val randomVal = (0..16).random()
      if (c.value == "x") {
        randomVal.toString(16)
      } else {
        ((randomVal and 0x3) or 0x8).toString(16)
      }
    }
  }

  private fun calcMd5(requestBody: String?): String {
    return if (requestBody == null) {
      ""
    } else {
      val md = MessageDigest.getInstance("MD5")
      md.update(requestBody.toByteArray())
      val digest = md.digest()
      Base64.getEncoder().encodeToString(digest)
    }
  }

  private fun urlToSign(queryParam: Map<String, String>?): String {
    return if (queryParam == null) {
      path
    } else {
      val sortedQuery = queryParam.toSortedMap()
      var queryString = ""
      for ((key, value) in sortedQuery) {
        val subQueryString = "$key=$value"
        queryString = if (queryString != "") "$queryString&$subQueryString" else subQueryString
      }
      "$path?$queryString"
    }
  }

  private fun getSign(): String {
    val headersToSign = mapOf(
      "X-Ca-Key" to System.getenv("APP_KEY"),
      "X-Ca-Nonce" to nonce,
      "X-Ca-SignatureMethod" to "HmacSHA256",
    )
    var textToSign = "$method\n*/*\n$md5\napplication/json\n$date\n"
    for ((key, value) in headersToSign) {
      textToSign += "$key:$value\n"
    }
    textToSign += url
    val hMacSHA256 = Mac.getInstance("HmacSHA256")
    val secretKey = SecretKeySpec(System.getenv("APP_SECRET").toByteArray(), "HmacSHA256")
    hMacSHA256.init(secretKey)
    val data = hMacSHA256.doFinal(textToSign.toByteArray())
    return Base64.getEncoder().encodeToString(data)
  }
}

File: cypress-test-code/cypress/utils/RequestAuth.ts
import * as CryptoJS from "crypto-js";

export default class RequestAuth {
  //cypress's default accept and content type, so ignore the validation of other content type in this test
  readonly accept = "*/*";
  readonly contentType = "application/json";
  readonly appKey = Cypress.env("appKey");
  readonly appSecret = Cypress.env("appSecret");

  method: "GET" | "POST" | "PUT" | "DELETE";
  path: string;
  queries: object;
  url: string;
  requestBody: object;
  requestHeader: object;

  constructor(
    method: "GET" | "POST" | "PUT" | "DELETE",
    path: string,
    queries: object,
    requestBody: object,
  ) {
    this.method = method;
    this.path = path;
    this.queries = queries;
    this.requestBody = requestBody;
    this.url = this.urlToSign();
    this.requestHeader = this.getRequestHeader()
  }

  private getRequestHeader() {
    let md5 = this.calcMd5();
    let date = new Date().toLocaleString();
    let nonce = this.createUuid();
    let textToSign =
      this.method + "\n" + this.accept + "\n" + md5 + "\n" + this.contentType + "\n" + date + "\n";
    let signatureHeaders;
    let headersToSign = new Map();
    //these three headers will be added to signature
    headersToSign.set("X-Ca-SignatureMethod", "HmacSHA256");
    headersToSign.set("X-Ca-Key", this.appKey);
    headersToSign.set("X-Ca-Nonce", nonce);
    let sortedKeys = Array.from(headersToSign.keys()).sort();
    for (let headerName of sortedKeys) {
      textToSign += headerName + ":" + headersToSign.get(headerName) + "\n";
      signatureHeaders = signatureHeaders ? signatureHeaders + "," + headerName : headerName;
    }
    //the url be added into signature should sort its queries first
    textToSign += this.urlToSign();
    let signature = CryptoJS.HmacSHA256(textToSign, this.appSecret).toString(CryptoJS.enc.Base64);
    //build the request header
    return {
      Date: date,
      "Content-MD5": md5,
      "X-Ca-Nonce": nonce,
      "X-Ca-Key": this.appKey,
      "X-Ca-Signature": signature,
      "X-Ca-SignatureMethod": "HmacSHA256",
      "X-Ca-Signature-Headers": signatureHeaders,
      "Content-Type": this.contentType,
    };
  }

  private urlToSign() {
    if (this.queries) {
      let sortedKeys = Object.keys(this.queries).sort();
      let queryString;
      for (let key of sortedKeys) {
        let s = key + "=" + this.queries[key];
        queryString = queryString ? queryString + "&" + s : s;
      }
      return this.path + "?" + queryString;
    } else {
      return this.path;
    }
  }

  private calcMd5() {
    //won't calculate MD5 if no request body or request body is empty
    if (this.requestBody && JSON.stringify(this.requestBody) !== "{}") {
      return CryptoJS.MD5(JSON.stringify(this.requestBody)).toString(CryptoJS.enc.Base64);
    } else {
      return "";
    }
  }

  private createUuid() {
    return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, function (c) {
      let r = (Math.random() * 16) | 0,
        v = c == "x" ? r : (r & 0x3) | 0x8;
      return v.toString(16);
    });
  }
}